tests/stencils/test_reduce.py

```diff
@@ -3,14 +3,56 @@
 
 
 def test_reduce():
+    transpile_and_validate(various_reductions)
     transpile_and_validate(kw_args)
 
 
-# TODO: Add tests for the following cases
-# different expressions
-# different location chains
-# nested reductions
-# expressions in weights (& init)
+@stencil
+def various_reductions(
+    vertex: Field[Vertex],
+    edge: Field[Edge, K],
+    cell: Field[Cell],
+    ve: Field[Vertex > Edge, K],
+    vc: Field[Vertex > Cell, K],
+    ev: Field[Edge > Vertex, K],
+    ec: Field[Edge > Cell, K],
+    cv: Field[Cell > Vertex, K],
+    ce: Field[Cell > Edge, K],
+    sparse1: Field[Vertex > Edge > Cell > Vertex > Edge > Cell],
+    sparse2: Field[Cell > Edge > Cell > Edge > Cell > Edge],
+    sparse3: Field[Cell > Vertex > Cell > Vertex > Cell > Edge],
+):
+    with levels_upward:
+        edge = sum_over(Edge > Cell, cell * ec)
+
+        cell = max_over(Cell > Vertex, pow(vertex, cv / cell))
+
+        vertex = reduce_over(
+            Vertex > Edge,
+            log(ve)
+            / max_over(
+                Edge > Cell > Vertex > Edge,
+                sin(edge[Edge]) / exp(edge[Edge > Cell > Vertex > Edge]) ** 5
+                - sum_over(
+                    Edge > Cell,
+                    sqrt(cell),
+                    weights=[
+                        edge[Edge],
+                        # FIXME: using `Edge > Cell > Vertex > Edge` as index segfaults
```

Comment:
So this is a limitation of dawn? I don't think this should be a problem from a theoretical point of view. Weights should be expressions in the _outer_ context whereas the reduction expression should be in the _inner_ context.

Anyways, I removed the FIXME. Should I open an issue in dawn?

Reply:
It's the current way dawn allows you to the express the outer context. Dawn is unaware of the surroundings when evaluating this sum reduction, so the offset must be expressed relatively to the current reduction. So locally it's `Edge` but we know that the absolute offset is actually `Edge > Cell > Vertex > Edge`.
Unless you want to open an issue about adding support for escaping the reduction's context, which means that offsets should be then interpreted as absolute rather than local, I'd say this representation is ok for the time being as long as it's well understood by us.
By the way, I suspect that the generated code in this case (`edge[Edge]`) could be incorrect, though. Have you checked?

Reply:
wait, so this would be _valid_ (no nested reductions/spare fills)?
```python
edge1: Field[Edge]
edge2: Field[Edge]
edge1 = sum_over(Edge > Cell > Edge, 1, weights = [
  edge2[Edge > Cell > Edge],
  ...
])
```
It shouldn't be, right? Since you said you disallow _inner_ accesses. But then you basically can't specify the horizontal offset for weights, since only one offset of the two boolean horizontal unstructured offsets are valid.

I haven't checked the generated code/SIR. Such checks are meant to be added later to dusk.

Reply:
Correct, that code is not valid, as inner accesses are disallowed within the reduction's weights.